server/errorstore/errors.go

```diff
@@ -0,0 +1,282 @@
+		}
+	} else if len(unpackedErr.ErrChain) > 0 {
+		lastFrame := unpackedErr.ErrChain[0].Frame
+		fmt.Fprintf(&sb, "%s:%d", lastFrame.File, lastFrame.Line)
+	}
+	return sha256b64(sb.String())
+}
+
+func hashAndMarshalError(externalErr error) (errHash string, errAsJson string, err error) {
+	m := eris.ToJSON(externalErr, true)
+	bytes, err := json.MarshalIndent(m, "", "  ")
+	if err != nil {
+		return "", "", err
+	}
+	return hashError(externalErr), string(bytes), nil
+}
+
+func (h *Handler) handleErrors(ctx context.Context) {
+	atomic.StoreInt32(&h.running, 1)
+	defer func() {
+		atomic.StoreInt32(&h.running, 0)
+	}()
+
+	if testOnStart != nil {
+		testOnStart()
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case err := <-h.errCh:
+			h.storeError(ctx, err)
+		}
+	}
+}
+
+func (h *Handler) storeError(ctx context.Context, err error) {
+	errorHash, errorJson, err := hashAndMarshalError(err)
+	if err != nil {
+		level.Error(h.logger).Log("err", err, "msg", "hashErr failed")
+		if testOnStore != nil {
+			testOnStore(err)
+		}
+		return
+	}
+	jsonKey := fmt.Sprintf("error:%s:json", errorHash)
+
+	conn := redis.ConfigureDoer(h.pool, h.pool.Get())
+	defer conn.Close()
+
+	secs := int(h.ttl.Seconds())
+	if secs <= 0 {
+		secs = 1 // SET EX fails if ttl is <= 0
+	}
+	if _, err := conn.Do("SET", jsonKey, errorJson, "EX", secs); err != nil {
+		level.Error(h.logger).Log("err", err, "msg", "redis SET failed")
+		if testOnStore != nil {
+			testOnStore(err)
+		}
+		return
+	}
+
+	if testOnStore != nil {
+		testOnStore(nil)
+	}
+}
+
+// Store handles the provided error by storing it into Redis if the handler is
+// still running. In any case, it always returns the error as provided.
+//
+// If the ctx is cancelled before the error is stored, the call returns without
+// storing the error, otherwise it waits for a predefined period of time to try
+// to store the error.
+func (h *Handler) Store(ctx context.Context, err error) error {
+	if atomic.LoadInt32(&h.running) == 0 {
+		return err
+	}
+
+	timer := time.NewTimer(2 * time.Second)
+	defer timer.Stop()
+	select {
+	case h.errCh <- err:
+	case <-timer.C:
+	case <-ctx.Done():
+	}
+
+	return err
+}
+
+// NewHttpHandler creates an http.HandlerFunc that flushes the errors stored
+// by the provided ErrorFlusher and returns them in the response as JSON.
+func NewHttpHandler(eh ErrorFlusher) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		errors, err := eh.Flush()
+		if err != nil {
+			w.WriteHeader(http.StatusBadGateway)
+			return
+		}
+
+		// each string returned by eh.Flush is already JSON-encoded, so to prevent
+		// double-marshaling while still marshaling the list of errors as a JSON
+		// array, treat them as raw json messages.
+		raw := make([]json.RawMessage, len(errors))
+		for i, s := range errors {
+			raw[i] = json.RawMessage(s)
+		}
+
+		bytes, err := json.Marshal(raw)
+		if err != nil {
+			w.WriteHeader(http.StatusBadGateway)
```

Comment:
This would be more of an internal server error, and also same as above, writing `err` could help troubleshooting.